### ISSUE TITLE
Trying to get it to work with SHTC3

### ISSUE THREAD

**Issue by lxne** [Tom-Hirschberger/MMM-Temperature#15]
Hi Tom,

I bought an [SHTC3](https://www.berrybase.de/adafruit-sensirion-shtc3-temperatur-feuchtigkeits-sensor) temperature and humidity sensor and found your module.

Unfortunately python is quite a mystery to me, but I was able to edit [shanewholloway's two scripts](https://gist.github.com/shanewholloway/cfb8e3d9013709c730aa24387cfd4a18) very slightly (shtc3.py and shtc3-data.py) which now result in:
`{'humidity': 47.701228, 'temperature_c': 23.13, 'temperature_f': 73.63, 'error': False}`

The shtc3-data.py file's content looks like this:

```
from __future__ import print_function
import time

from shtc3 import SHTC3

with SHTC3.with_bus_addr() as sensor:
    sensor.wake()
    reading = sensor.measure(clock=False)
    sensor.sleep()

    print(reading)
```

It depends on shtc3.py which I attached as [shtc3.txt](https://github.com/Tom-Hirschberger/MMM-Temperature/files/13114506/shtc3.txt). Right now I have absolutely no clue where "'error': False" should come from, so I just added it as a boolean. 

I moved both files into the script folder of MMM-Temperature and made sure they work there.

In the config.js I added…
```
       {
            module: "MMM-Temperature",
            position: "bottom_right",
            config: {
                temperatureHigh: 30,
                temperatureLow: 10,
                humidityHigh: 60,
                humidityLow: 30,
                            sensors: [
                    {
                        name: "Sensor",
                        script: "shtc3-data.py"
                    }
                ]
            }
        },
```

but it doesn't work.
The module won't show any values.

Now I am not sure where to start.
Formatting issue? Problem with the files? Anything else?
For me It's a bit like searching for the needle in the haystack.

It would be awesome if you could help me a bit.

Best wishes,
Alex

**Comment by Tom-Hirschberger** [comment]
Hi,
no problem. I try to help where i can...
I do not have a SHTC3 sensor so i could not test.

There are some things that might cause problems:
* I am not sure if imports of classes of files in the same directory work as i am not sure which directory is the current working directory -> Better put all code to one file
* The Shebang is missing (The first line in the script telling the Shell which interpreter to use) -> Add the Interpreter to the first line
* The Error field should be true if a error occurs during the reading of the values. Either cause the sensor is not reachable or anything else

I tried to add your scripts to one file and added the Shebang and some error handling. Maybe you give it a try:

```python3
#!/usr/bin/env python3
from __future__ import print_function
import time
import json
from contextlib import contextmanager
from smbus2 import SMBus, i2c_msg

class SHTC3:
    def __init__(self, bus, addr=0x70):
        self.bus = bus
        self.addr = addr

    def sleep(self):
        self._send_cmd(0xB098)

    def wake(self):
        self._send_cmd(0x3517)

    def reset(self):
        self._send_cmd(0x805d)

    def id(self):
        return self._send_cmd(0xefc8, 3)

    def measure(self, wait=0.05, **kw):
        if not callable(wait):
            wait = lambda td=wait: time.sleep(td)

        for ans in self.iter_measure(**kw):
            if ans is not None:
                return ans
            wait()

    def iter_measure(self, **kw):
        for raw in self._iter_raw_measure(**kw):
            if raw is not None:
                yield self._as_reading(raw)
            else: yield {"error": True}


    def _as_reading(self, raw) :

        h = round(raw['h'] / 655.35, 6) # Kommastelle von 65535. nach 655.35 verrückt
        temp_raw = round(raw['t'] / 65535., 6)
        return dict(humidity=h, # h zu humidity geändert
            temperature_c=self._temp_c(temp_raw), # t_c zu temperature_c geändert
            temperature_f=self._temp_f(temp_raw), # t_f zu temperature_c geändert
            error=bool(False)) # Pseudofehler fweil MMM-Temperature die Ausgabe erwartet
#            t_r=temp_raw)

    def _temp_f(self, temp_raw):
        return round(-49 + 315 * temp_raw, 2)

    def _temp_c(self, temp_raw):
        return round(-45 + 175 * temp_raw, 2)


    _measure_cmd = {
        ('normal', 'clock', 'th'):  0x7CA2,
        ('normal', 'clock', 'ht'):  0x5C24,
        ('normal', 'poll', 'th'): 0x7866,
        ('normal', 'poll', 'ht'): 0x58E0,

        ('low', 'clock', 'th'):  0x6458,
        ('low', 'clock', 'ht'):  0x44DE,
        ('low', 'poll', 'th'): 0x609C,
        ('low', 'poll', 'ht'): 0x401A,
    }

    def _iter_raw_measure(self, low_power=False, clock=False, humdity_first=False):
        cmd = self._measure_cmd[
            "low" if low_power else "normal",
            "clock" if clock else "poll",
            "ht" if humdity_first else "th"]

        try:
            self._send_cmd(cmd)

            reading = None
            while reading is None:
                reading = self._raw_reading(humdity_first)
                yield reading
        except:
            yield {"error": True}

    def _raw_reading(self, humdity_first):
        raw = i2c_msg.read(self.addr, 6)
        try: self.bus.i2c_rdwr(raw)
        except IOError:
            return None

        if humdity_first:
            [h_hi, h_low, h_crc, t_hi, t_low, t_crc] = raw
        else:
            [t_hi, t_low, t_crc, h_hi, h_low, h_crc] = raw

        t = (t_hi << 8) | t_low
        h = (h_hi << 8) | h_low
        return dict(t=t, h=h, t_crc=t_crc, h_crc=h_crc)


    def _send_cmd(self, cmd, len_read=None):
        c0 = (cmd >> 8) & 0xff
        c1 = (cmd >> 0) & 0xff
        mc = i2c_msg.write(self.addr, [c0,c1])

        if len_read is None:
            return self.bus.i2c_rdwr(mc)
        else:
            mr = i2c_msg.read(self.addr, len_read)
            self.bus.i2c_rdwr(mc, mr)
            return list(mr)

    @classmethod
    @contextmanager
    def with_bus_addr(klass, bus_id=0x01, addr=0x70):
        with SMBus(bus_id) as bus:
            yield klass(bus, addr)
            
with SHTC3.with_bus_addr() as sensor:
    try:
        sensor.wake()
        reading = sensor.measure(clock=False)
        sensor.sleep()
    except:
        reading = {"error": True}
    
    print(json.dumps(reading))
```

**Comment by lxne** [comment]
Thank you so much for your speedy reply.
I put the code in a file (shtc3) and I think we are almost there.

But when I run it from terminal with `python3 shtc3` I only get `{"error": true}`

If I comment out line 38 `            else: yield {"error": True}` and run it, I get an **error and the measurements**:
```
Exception ignored in: <generator object SHTC3._iter_raw_measure at 0x7f97f39200>
Traceback (most recent call last):
  File "/home/pi/MagicMirror/modules/MMM-Temperature/scripts/shtc3", line 37, in iter_measure
    yield self._as_reading(raw)
RuntimeError: generator ignored GeneratorExit
{"humidity": 48.821241, "temperature_c": 23.49, "temperature_f": 74.27, "error": false}
```

**Comment by Tom-Hirschberger** [comment]
Better but not the solution. Can you please test this version:

```python3
#!/usr/bin/env python3
from __future__ import print_function
import time
import json
from contextlib import contextmanager
from smbus2 import SMBus, i2c_msg

class SHTC3:
    def __init__(self, bus, addr=0x70):
        self.bus = bus
        self.addr = addr

    def sleep(self):
        self._send_cmd(0xB098)

    def wake(self):
        self._send_cmd(0x3517)

    def reset(self):
        self._send_cmd(0x805d)

    def id(self):
        return self._send_cmd(0xefc8, 3)

    def measure(self, wait=0.05, **kw):
        if not callable(wait):
            wait = lambda td=wait: time.sleep(td)

        for ans in self.iter_measure(**kw):
            if ans is not None:
                return ans
            wait()

    def iter_measure(self, **kw):
        for raw in self._iter_raw_measure(**kw):
            if raw is not None:
                yield self._as_reading(raw)
            else: None


    def _as_reading(self, raw) :

        h = round(raw['h'] / 655.35, 6) # Kommastelle von 65535. nach 655.35 verrückt
        temp_raw = round(raw['t'] / 65535., 6)
        return dict(humidity=h, # h zu humidity geändert
            temperature_c=self._temp_c(temp_raw), # t_c zu temperature_c geändert
            temperature_f=self._temp_f(temp_raw), # t_f zu temperature_c geändert
            error=bool(False)) # Pseudofehler fweil MMM-Temperature die Ausgabe erwartet
#            t_r=temp_raw)

    def _temp_f(self, temp_raw):
        return round(-49 + 315 * temp_raw, 2)

    def _temp_c(self, temp_raw):
        return round(-45 + 175 * temp_raw, 2)


    _measure_cmd = {
        ('normal', 'clock', 'th'):  0x7CA2,
        ('normal', 'clock', 'ht'):  0x5C24,
        ('normal', 'poll', 'th'): 0x7866,
        ('normal', 'poll', 'ht'): 0x58E0,

        ('low', 'clock', 'th'):  0x6458,
        ('low', 'clock', 'ht'):  0x44DE,
        ('low', 'poll', 'th'): 0x609C,
        ('low', 'poll', 'ht'): 0x401A,
    }

    def _iter_raw_measure(self, low_power=False, clock=False, humdity_first=False):
        cmd = self._measure_cmd[
            "low" if low_power else "normal",
            "clock" if clock else "poll",
            "ht" if humdity_first else "th"]

        try:
            self._send_cmd(cmd)

            reading = None
            while reading is None:
                reading = self._raw_reading(humdity_first)
                yield reading
        except:
            yield None

    def _raw_reading(self, humdity_first):
        raw = i2c_msg.read(self.addr, 6)
        try: self.bus.i2c_rdwr(raw)
        except IOError:
            return None

        if humdity_first:
            [h_hi, h_low, h_crc, t_hi, t_low, t_crc] = raw
        else:
            [t_hi, t_low, t_crc, h_hi, h_low, h_crc] = raw

        t = (t_hi << 8) | t_low
        h = (h_hi << 8) | h_low
        return dict(t=t, h=h, t_crc=t_crc, h_crc=h_crc)


    def _send_cmd(self, cmd, len_read=None):
        c0 = (cmd >> 8) & 0xff
        c1 = (cmd >> 0) & 0xff
        mc = i2c_msg.write(self.addr, [c0,c1])

        if len_read is None:
            return self.bus.i2c_rdwr(mc)
        else:
            mr = i2c_msg.read(self.addr, len_read)
            self.bus.i2c_rdwr(mc, mr)
            return list(mr)

    @classmethod
    @contextmanager
    def with_bus_addr(klass, bus_id=0x01, addr=0x70):
        with SMBus(bus_id) as bus:
            yield klass(bus, addr)
            
with SHTC3.with_bus_addr() as sensor:
    try:
        sensor.wake()
        reading = sensor.measure(clock=False)
        sensor.sleep()
    except:
        reading = {"error": True}
        
    if reading is None:
        reading = {"error": True}
    
    print(json.dumps(reading))
```

You should be able to run the script with:

```
./shtc3
```

This way the Shebang is used which in this case means something like search for `python3` in the environment `env` and use this binary as interpreter to run the code written in the lines after.

**Comment by Tom-Hirschberger** [comment]
I forgot to mention that i added a conversion to JSON in the first version of the script as the module expects a JSON as input and not a dump of a Python object. It looks the same but is different in many cases (like for boolean values).

**Comment by Tom-Hirschberger** [comment]
One more hint...
If you want to see the errors during the call in the script you can change the "except" in line 125 to "finally".
You than will see the exceptions.

**Comment by lxne** [comment]
Thank you.
I tried the code…
```
pi@raspberrypi:~/MagicMirror/modules/MMM-Temperature/scripts $ chmod +x shtc3
pi@raspberrypi:~/MagicMirror/modules/MMM-Temperature/scripts $ ./shtc3
{"error": true}
```
Unfortunately still the same result.
And 'finally' doesn't give me anything else in terminal. It stays the same `{"error": true}`
```
with SHTC3.with_bus_addr() as sensor:
    try:
        sensor.wake()
        reading = sensor.measure(clock=False)
        sensor.sleep()
    finally:
        reading = {"error": True}

    if reading is None:
        reading = {"error": True}

    print(json.dumps(reading))

```

**Comment by Tom-Hirschberger** [comment]
Sorry. Forgot a "yield"...

```python3
#!/usr/bin/env python3
from __future__ import print_function
import time
import json
from contextlib import contextmanager
from smbus2 import SMBus, i2c_msg

class SHTC3:
    def __init__(self, bus, addr=0x70):
        self.bus = bus
        self.addr = addr

    def sleep(self):
        self._send_cmd(0xB098)

    def wake(self):
        self._send_cmd(0x3517)

    def reset(self):
        self._send_cmd(0x805d)

    def id(self):
        return self._send_cmd(0xefc8, 3)

    def measure(self, wait=0.05, **kw):
        if not callable(wait):
            wait = lambda td=wait: time.sleep(td)

        for ans in self.iter_measure(**kw):
            if ans is not None:
                return ans
            wait()

    def iter_measure(self, **kw):
        for raw in self._iter_raw_measure(**kw):
            if raw is not None:
                yield self._as_reading(raw)
            else: yield None


    def _as_reading(self, raw) :

        h = round(raw['h'] / 655.35, 6) # Kommastelle von 65535. nach 655.35 verrückt
        temp_raw = round(raw['t'] / 65535., 6)
        return dict(humidity=h, # h zu humidity geändert
            temperature_c=self._temp_c(temp_raw), # t_c zu temperature_c geändert
            temperature_f=self._temp_f(temp_raw), # t_f zu temperature_c geändert
            error=bool(False)) # Pseudofehler fweil MMM-Temperature die Ausgabe erwartet
#            t_r=temp_raw)

    def _temp_f(self, temp_raw):
        return round(-49 + 315 * temp_raw, 2)

    def _temp_c(self, temp_raw):
        return round(-45 + 175 * temp_raw, 2)


    _measure_cmd = {
        ('normal', 'clock', 'th'):  0x7CA2,
        ('normal', 'clock', 'ht'):  0x5C24,
        ('normal', 'poll', 'th'): 0x7866,
        ('normal', 'poll', 'ht'): 0x58E0,

        ('low', 'clock', 'th'):  0x6458,
        ('low', 'clock', 'ht'):  0x44DE,
        ('low', 'poll', 'th'): 0x609C,
        ('low', 'poll', 'ht'): 0x401A,
    }

    def _iter_raw_measure(self, low_power=False, clock=False, humdity_first=False):
        cmd = self._measure_cmd[
            "low" if low_power else "normal",
            "clock" if clock else "poll",
            "ht" if humdity_first else "th"]

        try:
            self._send_cmd(cmd)

            reading = None
            while reading is None:
                reading = self._raw_reading(humdity_first)
                yield reading
        except:
            yield None

    def _raw_reading(self, humdity_first):
        raw = i2c_msg.read(self.addr, 6)
        try: self.bus.i2c_rdwr(raw)
        except IOError:
            return None

        if humdity_first:
            [h_hi, h_low, h_crc, t_hi, t_low, t_crc] = raw
        else:
            [t_hi, t_low, t_crc, h_hi, h_low, h_crc] = raw

        t = (t_hi << 8) | t_low
        h = (h_hi << 8) | h_low
        return dict(t=t, h=h, t_crc=t_crc, h_crc=h_crc)


    def _send_cmd(self, cmd, len_read=None):
        c0 = (cmd >> 8) & 0xff
        c1 = (cmd >> 0) & 0xff
        mc = i2c_msg.write(self.addr, [c0,c1])

        if len_read is None:
            return self.bus.i2c_rdwr(mc)
        else:
            mr = i2c_msg.read(self.addr, len_read)
            self.bus.i2c_rdwr(mc, mr)
            return list(mr)

    @classmethod
    @contextmanager
    def with_bus_addr(klass, bus_id=0x01, addr=0x70):
        with SMBus(bus_id) as bus:
            yield klass(bus, addr)
            
with SHTC3.with_bus_addr() as sensor:
    try:
        sensor.wake()
        reading = sensor.measure(clock=False)
        sensor.sleep()
    except:
        reading = {"error": True}
        
    if reading is None:
        reading = {"error": True}
    
    print(json.dumps(reading))
```

**Comment by lxne** [comment]
Oh boy, there was something else going on.
I now played around quite a bit.
Even asked chatGPT to help me dubug this :)
[Here you can take a look at the conversation.](https://chat.openai.com/share/898799f3-77e9-4822-a156-7de0fe434fa9)

After adding the debug print statements suggested by chatGPT, I got this…
```
Sensor connection established
{"error": true}
```
Somehow the script got no raw sensor data.

I went back to my two file setup and at first run it spat out errors too. But after running it again it worked!?
After some digging, I found out that somehow the script or sensor needs two wake calls.

So now I added another `sensor.wake()` after '#    print("Sensor connection established")'.
Now I get…
```
Sensor connection established
Raw sensor data: {'t': 25698, 'h': 31430, 't_crc': 142, 'h_crc': 181}
Sensor data received: {'humidity': 47.959106, 'temperature_c': 23.62, 'temperature_f': 74.52, 'error': False}
Exception ignored in: <generator object SHTC3._iter_raw_measure at 0x7fa03d0270>
Traceback (most recent call last):
  File "/home/pi/MagicMirror/modules/MMM-Temperature/scripts/./shtc3dev", line 39, in iter_measure
    yield self._as_reading(raw)
RuntimeError: generator ignored GeneratorExit
{"humidity": 47.959106, "temperature_c": 23.62, "temperature_f": 74.52, "error": false}
```

Here is the new code (debug statements commented out). Result below.
```python
#!/usr/bin/env python3
from __future__ import print_function
import time
import json
from contextlib import contextmanager
from smbus2 import SMBus, i2c_msg

class SHTC3:
    def __init__(self, bus, addr=0x70):
        self.bus = bus
        self.addr = addr

    def sleep(self):
        self._send_cmd(0xB098)

    def wake(self):
        self._send_cmd(0x3517)

    def reset(self):
        self._send_cmd(0x805d)

    def id(self):
        return self._send_cmd(0xefc8, 3)

    def measure(self, wait=0.05, **kw):
        if not callable(wait):
            wait = lambda td=wait: time.sleep(td)

        for ans in self.iter_measure(**kw):
            if ans is not None:
#                print("Sensor data received:", ans)
                return ans
            wait()

    def iter_measure(self, **kw):
        for raw in self._iter_raw_measure(**kw):
            if raw is not None:
#                print("Raw sensor data:", raw)
                yield self._as_reading(raw)
            else: yield None


    def _as_reading(self, raw) :

        h = round(raw['h'] / 655.35, 6) # Kommastelle von 65535. nach 655.35 verrückt
        temp_raw = round(raw['t'] / 65535., 6)
        return dict(humidity=h, # h zu humidity geändert
            temperature_c=self._temp_c(temp_raw), # t_c zu temperature_c geändert
            temperature_f=self._temp_f(temp_raw), # t_f zu temperature_c geändert
            error=bool(False)) # Pseudofehler weil MMM-Temperature die Ausgabe erwartet
#            t_r=temp_raw)

    def _temp_f(self, temp_raw):
        return round(-49 + 315 * temp_raw, 2)

    def _temp_c(self, temp_raw):
        return round(-45 + 175 * temp_raw, 2)


    _measure_cmd = {
        ('normal', 'clock', 'th'):  0x7CA2,
        ('normal', 'clock', 'ht'):  0x5C24,
        ('normal', 'poll', 'th'): 0x7866,
        ('normal', 'poll', 'ht'): 0x58E0,

        ('low', 'clock', 'th'):  0x6458,
        ('low', 'clock', 'ht'):  0x44DE,
        ('low', 'poll', 'th'): 0x609C,
        ('low', 'poll', 'ht'): 0x401A,
    }

    def _iter_raw_measure(self, low_power=False, clock=False, humdity_first=False):
        cmd = self._measure_cmd[
            "low" if low_power else "normal",
            "clock" if clock else "poll",
            "ht" if humdity_first else "th"]

        try:
            self._send_cmd(cmd)

            reading = None
            while reading is None:
                reading = self._raw_reading(humdity_first)
                yield reading
        except:
            yield None

    def _raw_reading(self, humdity_first):
        raw = i2c_msg.read(self.addr, 6)
        try: self.bus.i2c_rdwr(raw)
        except IOError:
            return None

        if humdity_first:
            [h_hi, h_low, h_crc, t_hi, t_low, t_crc] = raw
        else:
            [t_hi, t_low, t_crc, h_hi, h_low, h_crc] = raw

        t = (t_hi << 8) | t_low
        h = (h_hi << 8) | h_low
        return dict(t=t, h=h, t_crc=t_crc, h_crc=h_crc)


    def _send_cmd(self, cmd, len_read=None):
        c0 = (cmd >> 8) & 0xff
        c1 = (cmd >> 0) & 0xff
        mc = i2c_msg.write(self.addr, [c0,c1])

        if len_read is None:
            return self.bus.i2c_rdwr(mc)
        else:
            mr = i2c_msg.read(self.addr, len_read)
            self.bus.i2c_rdwr(mc, mr)
            return list(mr)

    @classmethod
    @contextmanager
    def with_bus_addr(klass, bus_id=0x01, addr=0x70):
        with SMBus(bus_id) as bus:
            yield klass(bus, addr)

with SHTC3.with_bus_addr() as sensor:
#    print("Sensor connection established")
    sensor.wake()
    try:
        sensor.wake()
        reading = sensor.measure(clock=False)
        sensor.sleep()
#    except Exception as e:
    except:
#        print("An error occured:", e)
        reading = {"error": True}

    if reading is None:
        reading = {"error": True}

    print(json.dumps(reading))
```

Results in:
```
Exception ignored in: <generator object SHTC3._iter_raw_measure at 0x7f8b111200>
Traceback (most recent call last):
  File "/home/pi/MagicMirror/modules/MMM-Temperature/scripts/./shtc3dev", line 39, in iter_measure
    yield self._as_reading(raw)
RuntimeError: generator ignored GeneratorExit
{"humidity": 47.685969, "temperature_c": 23.56, "temperature_f": 74.41, "error": false}
```

There is still something going on, but the sensor data is there.
Is it doing something that doesn't need to be done?

**Comment by Tom-Hirschberger** [comment]
It looks like the problem is the yield None in the _iter_raw_measure function might be the problem (see https://stackoverflow.com/questions/27755989/generatorexit-in-while-loop).
I do not have access to my laptop at the mobile but it think some smaller try-except blocks in the function instead of the one big block might solve the problem.

**Comment by lxne** [comment]
Puh, that seems to be a view steps above my almost not existing level.

ChatGPT gave me…
```python
def iter_measure(self, **kw):
    results = []
    for raw in self._iter_raw_measure(**kw):
        if raw is not None:
            results.append(self._as_reading(raw))
        else:
            results.append(None)
    return results
```

This really works without an error but it takes more than three seconds, while the other solution with error gets to the results almost instantly.
```
pi@raspberrypi:~/MagicMirror/modules/MMM-Temperature/scripts $ time ./shtc3dev
{"humidity": 46.814679, "temperature_c": 23.54, "temperature_f": 74.38, "error": false}

real	0m3,294s
user	0m0,096s
sys	0m0,031s
```

**Comment by Tom-Hirschberger** [comment]
Does it work without error if you use the original two files?
It might be easier to convert the result object in that case.

**Comment by Tom-Hirschberger** [comment]
OK. I think the error handling I added is crap. Please try this one:

```python3
#!/usr/bin/env python3
from __future__ import print_function
import time
import json
from contextlib import contextmanager
from smbus2 import SMBus, i2c_msg

class SHTC3:
    def __init__(self, bus, addr=0x70):
        self.bus = bus
        self.addr = addr

    def sleep(self):
        self._send_cmd(0xB098)

    def wake(self):
        self._send_cmd(0x3517)

    def reset(self):
        self._send_cmd(0x805d)

    def id(self):
        return self._send_cmd(0xefc8, 3)

    def measure(self, wait=0.05, **kw):
        if not callable(wait):
            wait = lambda td=wait: time.sleep(td)

        for ans in self.iter_measure(**kw):
            if ans is not None:
                return ans
            wait()

    def iter_measure(self, **kw):
        for raw in self._iter_raw_measure(**kw):
            if raw is not None:
                yield self._as_reading(raw)
            else: yield None


    def _as_reading(self, raw) :
        h = round(raw['h'] / 65535., 6)
        temp_raw = round(raw['t'] / 65535., 6)
        return dict(h=h,
            t_f=self._temp_f(temp_raw),
            t_c=self._temp_c(temp_raw),
            t_r=temp_raw)

    def _temp_f(self, temp_raw):
        return round(-49 + 315 * temp_raw, 2)

    def _temp_c(self, temp_raw):
        return round(-45 + 175 * temp_raw, 2)


    _measure_cmd = {
        ('normal', 'clock', 'th'):  0x7CA2,
        ('normal', 'clock', 'ht'):  0x5C24,
        ('normal', 'poll', 'th'): 0x7866,
        ('normal', 'poll', 'ht'): 0x58E0,

        ('low', 'clock', 'th'):  0x6458,
        ('low', 'clock', 'ht'):  0x44DE,
        ('low', 'poll', 'th'): 0x609C,
        ('low', 'poll', 'ht'): 0x401A,
    }

    def _iter_raw_measure(self, low_power=False, clock=False, humdity_first=False):
        cmd = self._measure_cmd[
            "low" if low_power else "normal",
            "clock" if clock else "poll",
            "ht" if humdity_first else "th"]

        self._send_cmd(cmd)

        reading = None
        while reading is None:
            reading = self._raw_reading(humdity_first)
            yield reading

    def _raw_reading(self, humdity_first):
        raw = i2c_msg.read(self.addr, 6)
        try: self.bus.i2c_rdwr(raw)
        except IOError:
            return None

        if humdity_first:
            [h_hi, h_low, h_crc, t_hi, t_low, t_crc] = raw
        else:
            [t_hi, t_low, t_crc, h_hi, h_low, h_crc] = raw

        t = (t_hi << 8) | t_low
        h = (h_hi << 8) | h_low
        return dict(t=t, h=h, t_crc=t_crc, h_crc=h_crc)


    def _send_cmd(self, cmd, len_read=None):
        c0 = (cmd >> 8) & 0xff
        c1 = (cmd >> 0) & 0xff
        mc = i2c_msg.write(self.addr, [c0,c1])

        if len_read is None:
            return self.bus.i2c_rdwr(mc)
        else:
            mr = i2c_msg.read(self.addr, len_read)
            self.bus.i2c_rdwr(mc, mr)
            return list(mr)

    @classmethod
    @contextmanager
    def with_bus_addr(klass, bus_id=0x01, addr=0x70):
        with SMBus(bus_id) as bus:
            yield klass(bus, addr)

with SHTC3.with_bus_addr() as sensor:
    sensor.wake()
    sensor.wake()
    reading = sensor.measure(clock=False)
    sensor.sleep()

    print(json.dumps(reading))
```

**Comment by lxne** [comment]
This works.
And it also works if the last part looks like…
```python
with SHTC3.with_bus_addr() as sensor:
    sensor.wake()
    try:
        sensor.wake()
        reading = sensor.measure(clock=False)
        sensor.sleep()
    except:
        reading = {"error": True}

    if reading is None:
        reading = {"error": True}

    print(json.dumps(reading))
```

**Comment by lxne** [comment]
Module gets the values 💯 
Thank you so much!

This is the code I use now:

```python
#!/usr/bin/env python3
from __future__ import print_function
import time
import json
from contextlib import contextmanager
from smbus2 import SMBus, i2c_msg


class SHTC3:
    def __init__(self, bus, addr=0x70):
        self.bus = bus
        self.addr = addr

    def sleep(self):
        self._send_cmd(0xB098)

    def wake(self):
        self._send_cmd(0x3517)

    def reset(self):
        self._send_cmd(0x805d)

    def id(self):
        return self._send_cmd(0xefc8, 3)

    def measure(self, wait=0.05, **kw):
        if not callable(wait):
            wait = lambda td=wait: time.sleep(td)

        for ans in self.iter_measure(**kw):
            if ans is not None:
                return ans
            wait()

    def iter_measure(self, **kw):
        for raw in self._iter_raw_measure(**kw):
            if raw is not None:
                yield self._as_reading(raw)
            else: yield None


    def _as_reading(self, raw) :

        h = round(raw['h'] / 655.35, 6) # Kommastelle von 65535. nach 655.35 verrückt
        temp_raw = round(raw['t'] / 65535., 6)
        return dict(humidity=h, # h zu humidity geändert
            temperature_c=self._temp_c(temp_raw), # t_c zu temperature_c geändert
            temperature_f=self._temp_f(temp_raw), # t_f zu temperature_c geändert
            error=bool(False)) # Ohne Funktion, nur weil MMM-Temperature die Ausgabe erwartet
#            t_r=temp_raw)

    def _temp_f(self, temp_raw):
        return round(-49 + 315 * temp_raw, 2)

    def _temp_c(self, temp_raw):
        return round(-45 + 175 * temp_raw, 2)


    _measure_cmd = {
        ('normal', 'clock', 'th'):  0x7CA2,
        ('normal', 'clock', 'ht'):  0x5C24,
        ('normal', 'poll', 'th'): 0x7866,
        ('normal', 'poll', 'ht'): 0x58E0,

        ('low', 'clock', 'th'):  0x6458,
        ('low', 'clock', 'ht'):  0x44DE,
        ('low', 'poll', 'th'): 0x609C,
        ('low', 'poll', 'ht'): 0x401A,
    }

    def _iter_raw_measure(self, low_power=False, clock=False, humdity_first=False):
        cmd = self._measure_cmd[
            "low" if low_power else "normal",
            "clock" if clock else "poll",
            "ht" if humdity_first else "th"]

        self._send_cmd(cmd)

        reading = None
        while reading is None:
            reading = self._raw_reading(humdity_first)
            yield reading

    def _raw_reading(self, humdity_first):
        raw = i2c_msg.read(self.addr, 6)
        try: self.bus.i2c_rdwr(raw)
        except IOError:
            return None

        if humdity_first:
            [h_hi, h_low, h_crc, t_hi, t_low, t_crc] = raw
        else:
            [t_hi, t_low, t_crc, h_hi, h_low, h_crc] = raw

        t = (t_hi << 8) | t_low
        h = (h_hi << 8) | h_low
        return dict(t=t, h=h, t_crc=t_crc, h_crc=h_crc)


    def _send_cmd(self, cmd, len_read=None):
        c0 = (cmd >> 8) & 0xff
        c1 = (cmd >> 0) & 0xff
        mc = i2c_msg.write(self.addr, [c0,c1])

        if len_read is None:
            return self.bus.i2c_rdwr(mc)
        else:
            mr = i2c_msg.read(self.addr, len_read)
            self.bus.i2c_rdwr(mc, mr)
            return list(mr)

    @classmethod
    @contextmanager
    def with_bus_addr(klass, bus_id=0x01, addr=0x70):
        with SMBus(bus_id) as bus:
            yield klass(bus, addr)

with SHTC3.with_bus_addr() as sensor:
    sensor.wake()
    try:
        sensor.wake()
        reading = sensor.measure(clock=False)
        sensor.sleep()
    except:
        reading = {"error": True}

    if reading is None:
        reading = {"error": True}

    print(json.dumps(reading))

```

You might wanna add it to MMM-Temperature's scripts folder?
Thanks again for helping me out.

**Comment by Tom-Hirschberger** [comment]
Great to hear that. I wanted to ask if I can add it to the next release. So yes I will add it to the scripts folder and add it to the configuration examples.

**Comment by lxne** [comment]
Hi Tom, I just did a new MM installation. That's when I saw that SHTC3 isn't yet in the scripts' folder.
Attached you can find the script. After more than half a year of use, I would say it has been thoroughly tested :)
[shtc3.zip](https://github.com/user-attachments/files/15846344/shtc3.zip)

**Comment by Tom-Hirschberger** [comment]
Hi and sorry. Do not know how i could have missed that.
Just released 0.0.9 which contains your version of the script a few seconds ago.

**Comment by lxne** [comment]
No worries.
Thank you!
